src/client/Client.cc

```diff
@@ -5309,24 +5309,48 @@ void Client::handle_caps(const MConstRef<MClientCaps>& m)
 
   got_mds_push(session.get());
 
+  bool do_cap_release = false;
   Inode *in;
   vinodeno_t vino(m->get_ino(), CEPH_NOSNAP);
   if (auto it = inode_map.find(vino); it != inode_map.end()) {
     in = it->second;
+
+    /* MDS maybe waiting for cap release with increased seq */
+    switch (m->get_op()) {
+      case CEPH_CAP_OP_REVOKE:
+      case CEPH_CAP_OP_GRANT:
+        if (!in->caps.count(mds)) {
+	  do_cap_release = true;
+          ldout(cct, 5) << __func__ << " vino " << vino << " don't have cap "
+                        << m->get_cap_id() << " op " << m->get_op()
+                        << ", immediately releasing" << dendl;
+	}
+    }
   } else {
-    if (m->get_op() == CEPH_CAP_OP_IMPORT) {
-      ldout(cct, 5) << __func__ << " don't have vino " << vino << " on IMPORT, immediately releasing" << dendl;
-      session->enqueue_cap_release(
-        m->get_ino(),
-        m->get_cap_id(),
-        m->get_seq(),
-        m->get_mseq(),
-        cap_epoch_barrier);
-    } else {
-      ldout(cct, 5) << __func__ << " don't have vino " << vino << ", dropping" << dendl;
```

Comment:
The `else` case flushes the queued release caps. This behaviour is modified by this change - do you see any issues due to that?

Reply:
No, the `else` case will only print this debug log and the flush just out side the `else` case.

Reply:
Right, but the current code will always flush queued cap releases if vino isn't found.

Reply:
IMO that's should be okay and the flush cap release will be trigger per tick.

Why it flushed it here is because the `MDS` may waiting for the cap release here in `IMPORT` case. And only when the MDS is waiting for it will it make sense to trigger the flush immediately here, or the MDS may will be stuck for 1 seconds at most.

Reply:
ACK. Fair enough.